src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/events/ManageExternalAppointmentsService.kt

```diff
@@ -148,66 +134,120 @@ class ManageExternalAppointmentsService(
+
+  // This should never happen but if it ever happens we are throwing NPE with a bit more context to it!
+  private fun PrisonAppointment.bookingId() = prisonerSearchClient.getPrisoner(prisonerNumber)?.bookingId?.toLong()
+    ?: throw NullPointerException("EXTERNAL APPOINTMENTS: Booking id not found for prisoner $prisonerNumber for prison appointment $prisonAppointmentId")
+
+  private fun BookingHistoryAppointment.internalLocationId() = nomisMappingClient.getNomisLocationMappingBy(prisonLocationId)?.nomisLocationId
+    ?: throw NullPointerException("EXTERNAL APPOINTMENTS: Internal location id for key $prisonLocationId not found for prison appointment ${bookingHistoryAppointmentId}Id")
+}
+
+private class NomisMatcher(
+  private val prisonApiClient: PrisonApiClient,
+  private val supportedAppointmentTypes: SupportedAppointmentTypes,
+) {
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
   }
 
-  private fun Collection<AppointmentSearchResult>.findMatchingAppointments(bha: BookingHistoryAppointment): List<AppointmentSearchResult> = filter { searchResult ->
-    searchResult.isTheSameTime(bha) && searchResult.isTheSameAppointmentType(supportedAppointmentTypes.typeOf(bha.bookingType())) && searchResult.isCancelled.not()
-  }.ifEmpty {
-    emptyList<AppointmentSearchResult>()
-      .also { log.info("EXTERNAL APPOINTMENTS: no matching appointments found in A&A for booking history appointment ${bha.bookingHistoryAppointmentId}") }
+  fun findMatchingAppointments(appointment: PrisonAppointment, internalLocationId: Long): Collection<Long> = run {
+    prisonApiClient.getPrisonersAppointmentsAtLocations(
+      prisonCode = appointment.prisonCode(),
+      prisonerNumber = appointment.prisonerNumber,
+      onDate = appointment.appointmentDate,
+      internalLocationId,
+    ).findMatching(appointment).map { it.eventId }
+  }
+
+  fun findMatchingAppointments(appointment: BookingHistoryAppointment, internalLocationId: Long): Collection<Long> = run {
+    prisonApiClient.getPrisonersAppointmentsAtLocations(
+      prisonCode = appointment.prisonCode,
+      prisonerNumber = appointment.prisonerNumber,
+      onDate = appointment.appointmentDate,
+      internalLocationId,
+    ).findMatching(appointment).map { it.eventId }
   }
 
-  private fun Collection<PrisonerSchedule>.findMatchingPrisonApi(appointment: PrisonAppointment): List<PrisonerSchedule> = filter { schedule ->
-    schedule.isTheSameTime(appointment) && schedule.isTheSameAppointmentType(supportedAppointmentTypes.typeOf(appointment.bookingType()))
+  private fun Collection<PrisonerSchedule>.findMatching(appointment: PrisonAppointment): List<PrisonerSchedule> = filter { schedule ->
+    schedule.isTheSameTime(appointment) &&
+      (schedule.isTheSameAppointmentType(supportedAppointmentTypes.typeOf(appointment.bookingType())) || supportedAppointmentTypes.isSupported(schedule.appointmentCode()))
```

Comment:
The code after`||` on this line is the key change here.  Sames for those that follow.